cupyx/scipy/sparse/csr.py

```diff
@@ -357,3 +372,223 @@ def isspmatrix_csr(x):
 
     """
     return isinstance(x, csr_matrix)
+
+
+def multiply_by_scalar(sp, a):
+    data = sp.data * a
+    indices = sp.indices.copy()
+    indptr = sp.indptr.copy()
+    return csr_matrix((data, indices, indptr), shape=sp.shape)
+
+
+def multiply_by_dense(sp, dn):
+    sp_m, sp_n = sp.shape
+    dn_m, dn_n = dn.shape
+    if not (sp_m == dn_m or sp_m == 1 or dn_m == 1):
+        raise ValueError('inconsistent shape')
+    if not (sp_n == dn_n or sp_n == 1 or dn_n == 1):
+        raise ValueError('inconsistent shape')
+    m, n = max(sp_m, dn_m), max(sp_n, dn_n)
+    nnz = sp.nnz * (m // sp_m) * (n // sp_n)
+    dtype = functools.reduce(numpy.promote_types, (sp.dtype, dn.dtype))
+    data = cupy.empty(nnz, dtype=dtype)
+    indices = cupy.empty(nnz, dtype=sp.indices.dtype)
+    if m > sp_m:
+        if n > sp_n:
+            indptr = cupy.arange(0, nnz+1, n, dtype=sp.indptr.dtype)
+        else:
+            indptr = cupy.arange(0, nnz+1, sp.nnz, dtype=sp.indptr.dtype)
+    else:
+        indptr = sp.indptr.copy()
+        if n > sp_n:
+            indptr *= n
+
+    # out = sp * dn
+    cupy_multiply_by_dense()(sp.data, sp.indptr, sp.indices, sp_m, sp_n,
+                             dn, dn_m, dn_n, indptr, m, n, data, indices)
+
+    return csr_matrix((data, indices, indptr), shape=(m, n))
+
+
+def cupy_multiply_by_dense():
```

Comment:
please memoize this per device
`@cupy.util.memoize(for_each_device=True)`

Reply:
OK, will add it.